### ISSUE TITLE
Macro problem: Search fails, rest of macro executes regardless

### ISSUE THREAD

**Issue by Rekrullurker** [notepad-plus-plus/notepad-plus-plus#4809]
### Description of the Issue
If you record a macro that contains a search, and that search fails to find the specified text, the rest of the macro will execute regardless. This makes it dangerous to use any macro that contains a search, especially if you use the option to repeat the macro to the end of the file.

### Steps to Reproduce the Issue
1. Make three lines that say: This is a *test* more stuff
2. Make three lines that say :This is a test more stuff
3. Record a macro that searches for an asterisk, moves to the right of the asterisk and deletes from there to the beginning of the line. Then search for the asterisk again, move to the left one space, delete it, then move to the start of the line and move down one line. Stop recording. Now run the macro to the end of the file.

### Expected Behavior
The macro should only execute on the lines containing asterisks and you should end up with;

test more stuff
test more stuff
test more stuff
This is a test more stuff
This is a test more stuff
This is a test more stuff

### Actual Behavior
Instead you get this;

test more stuff
test more stuff
test more stuffhis is a test more stuffhis is a test more stuffhis is a test more stuff

The macro works fine on the lines with asterisks, but when it comes to the final three lines, the search fails so it just executes the rest of the macro anyway. Also, I'm not exactly clear on why the remaining lines ended up appended to the end of the third line, given that none of the deletes were at the end of a line and the backspace key was never used. Somehow it seems to have deleted the EOF characters.

You can create some very complex and powerful macros by using the search feature to locate parts of the file that you want to change, but it's pretty much rendered useless if it's just going to ignore the fact that the required text wasn't found.

**Comment by ClaudiaFrank** [comment]
Imho a macro is dump it takes what it got and executes it.
If you need logic then you need some kind of scripting language and
there are plenty available as plugins using languages like python, lua, c#, java, ...
From my point of view this issue should be closed.

**Comment by Rekrullurker** [comment]
Then what's the point of being able to include a search in a macro if it doesn't matter whether the search succeeds or not? If you can't reliably use the search function in a macro to only change the parts of a file that match the search terms, then there's no point to even being able to include it in a macro. More than that, it's dangerous in that novice users may unwitting corrupt their files by running such a macro on the entire file without realizing that it will continue blindly making changes after the last occurrence of the search term is found.

Can you please provide an example of a situation where you would want to use search (not search & replace) in a macro, but **NOT** have the rest of the macro be conditional on the search succeeding? Conversely, can you provide an example of how having a macro terminate when a search (not search & replace) fails to find the specified terms would break something?

I'm not asking for the macro function to be made into a scripting language of its own with user-defined conditional execution and the like, I simply think it should at least be an option to have the macro terminate if the search fails. If the user searched for something, it stands to reason that they want the macro to find what they specified before taking whatever other actions they added. If the search fails, then the rest of the actions recorded in the macro aren't needed.

None of the scripting languages you mentioned are anywhere as easy to learn or use as simply having the program to record your keystrokes. Complex macros can easily be recorded by complete novices, but it falls apart if those macros only work under certain conditions. Telling people that they should learn a scripting language from scratch because you don't think a macro function should have any intelligence to it, smacks of elitism.

**Comment by ClaudiaFrank** [comment]
If you want to have a macro to behave more intelligent, than you have to define its behavior which means
you need to describe how to use it and under which circumstances - in code. 
In this context, assuming that one wants to execute the macro in documents which do NOT contain 
a particular text is as valid as one wants to run the macro if it HAS the text looking for.
How could we know beforehand? 

Concerning learning new things - how could we know what is easy for one and what isn't?
I'm pretty confused about your **smacks of elitism** remark - I do not know where you are coming from
but as long as you have access to the internet EVERYONE has access to mostly anything you can think of.

**Comment by sasumner** [comment]
I think macros in Notepad++ are supposed to be very simple things (by design).  I often record searches or replacements as macros, and sometimes chain together several in one macro, but I don't write them when a subsequent action relies on an earlier one.  I agree with Claudia in that that's better suited to scripting.

**Comment by Rekrullurker** [comment]
To clear up any possible confusion, I am **_NOT_** suggesting that the behavior of macros in regard to Search **_AND_** Replace be changed in any way. "Search" (technically "Find") is a completely separate menu item from "Replace", which initiates a Search **_AND_** Replace. I'm not trying to break everyone's existing macros. In the message below I am speaking exclusively about the stand-alone Find function which is invoked by CTRL-F. At no point should any of my comments be construed to apply to the Replace function which is invoked by pressing CTRL-H. As they are separate commands, a macro can easily differentiate between the two.

When you use Find, it will reposition the cursor if the search text is found. In fact, that is the sole purpose of the Find command. There is no other reason to use it. If it is not found, the cursor doesn't move. In normal editing, Find is used to find parts of a document that you want to make changes to. If you search for something and it's not found, do you just go ahead and make the planned changes at the current cursor position regardless? Of course not. So why would you ever want a macro to behave this way?

Since the Find command does nothing else by itself other than reposition the cursor if the search text is found, there is no conceivable reason anyone would ever use it in a macro. In other words, the sole purpose of the Find command is to conditionally reposition the cursor to a specific point in a file for the sole purpose of making changes to that specific part of the file.

Can you give me an honest, actual example of a situation where you would use Find in a macro, but want the macro to execute regardless of whether the cursor was repositioned to another part of the file _OR_ whether it simply stayed at its current position? Both of these conditions would produce wildly different results and there would be no reason for a user to ever create a macro that doesn't work reliably in every situation. Conversely, can you give me an honest, actual example of how it would break something if the macro terminates early when a Find command fails?

Again, this isn't Replace I'm talking about, but the separate Find command which only repositions the cursor. If a user wants changes made at the current cursor position, they simply record them there. If they use the Find command, it means they want the cursor repositioned to a specific part of the file.

Frankly, I never thought this would be a controversial idea. As it stands now, almost nobody would ever use Find in a macro unless they're 100% sure that at least the last line of the file contains the search text, otherwise it's going to blindly continue making changes to the file after the last instance of the search text is found. That severely limits the use of the Find command in Macros. Consequently, if nobody is using the Find command in macros, the change I suggested (stop execution of the macro if Find fails) would not affect anyone's existing macros. Even more so if it's a user-selectable option in prefs. 

It shouldn't even break the scripts of people who have used Find in a macro. Because the only time it can be reliably used is if they know for a fact that the last line of the file contains their search text, the macro would still work just fine because the termination condition would never happen. It would reach the end of the file and stop, just as its supposed to. 

In other words you couldn't use my example macro on this;

This is a XtestX more stuff
This is a XtestX more stuff
This is a test more stuff
This is a test more stuff

Because it will mangle the last two lines. You could only use it if the file looked like this;

This is a XtestX more stuff
This is a test more stuff
This is a test more stuff
This is a XtestX more stuff

It would skip over the middle lines, change the last line and exit normally. If you made macro execution conditional on the search text being found, the macro would still perform as expected because it would reach the end of the file without ever encountering the termination condition. However, it would also work just fine on my first example because it would change the first two lines, see that there are no more instances of the matching text and exit. Nothing gets broken because the very limited circumstance where search is used in a macro would already be situations where the user has made sure that the macro execution will only happen at specific points of the file. Making it conditional just takes that burden away from the user.

My comments about elitism were in regard to the fact that you would prefer users to learn a scripting language rather than being able to make full use of the much simpler macro recording facility already built into the program. While it's true you can't know what is or isn't easy for a person to learn, you can't deny that learning a scripting language is more complex than just recording a macro. Macros are intuitive enough that even novice users can understand the process. On the other hand, learning a scripting language can be a longer, more involved process. You can argue that the rewards are much greater, but most people don't want to learn a whole language just to be able to create macros, especially when the program already has its own macro recorder. Trust me, there are people for whom unpacking a Rar file requires itemized instructions and a video tutorial, every time they do it. Such people have no hope of learning scripting. 

To be honest, I don't know any of those languages and don't really feel inclined to learn. I've written some Windows batch files and generally found it to be a colossal pain in the a** to include any kind of logic in them due to Microsoft's a**-backward variable handling where you have to jump through hoops to ensure that variables actually behave the way logic tells you that they should. Not to mention trying to keep straight all the various ways variables have to be referenced depending on what type of variable it is or what you're doing with it. I usually end up spending a minimum of half an hour trying to figure out what obscure condition is preventing my perfectly logical script from working properly. Sometime I have to ask in a forum so that someone can explain why you need to do something completely different because you used a different command. I've still never been able to get the errorlevel variable to work reliably inside a loop to determine if a command executed successfully or not. For that matter, I've never understood why you have to test the errorlevel command in reverse order from highest possible value to lowest, rather than it having an absolute value that you can test for. This kind of crap drives me nuts.

Again, I'm not trying to break anyone's existing scripts. The only conceivable situations where you would ever use Find in macro are situations where you have already manually made sure it's not going to blindly make changes to the file. Making the macro execution conditional just automates that process.

**Comment by deadeyelo** [comment]
To ClaudiaFrank - My $0.02 - You might want to rethink your argument on this.  Of the several editors that support macros that I have worked with, Windows, OpenVMS, they abort macro execution on some errors failed search, down arrow, up arrow when at end or beginning of buffer, execution for more iterations than exist in file (the Run Multiple Times, executing 10 times when only 5 times is possible),  etc.  The TECO editor has searches that allow 'if success then else fail' as an option to searches (one can issue a search that doesn't check for failure, and when done in an iteration corrupts really good).

I don't do much with UNIX/LINUX so I don't know for sure editors there abort on an error in a macro, but would guess that they do.   I also am not proficient enough with UNIX/LINUX editors to easily test them for this.  May give it a try.

I am not sure but I think the 32 bit version of NotePad++ aborted on search fail or cursor movement that couldn't be done (line, word, character movement, not sure about begin of line, end of line if already there) did abort.  I changed to 64 bit earlier last year and felt 'surprised' when I first noticed that execution continued after and 'error' occurred.  I am thinking of going back to 32 bit as 64 bit lost several Plug-Ins I used and MACRO's don't execute reliably (don't remember having that problem until it showed up one day).  If I do that I will post an update.

**Comment by Arkaine80** [comment]
TLDR: My suggested "fix" is to extend the Run Macro Multiple times ... to include an option to "Run until find fails".

This is an old post, but I came into this issue myself today (x64). 
I am trying to extract a specific element in a JSON array. 

I start by finding the main tag let's go with "Data":, do a find for that element let's go with Fullname, move right until after the ":, shift End -2 chars, cut, ctrl+home.

I then use Run this multiple times, but I have to know the exact amount to run this, or it starts cutting the wrong data.

The screen "freezes" and I need to wait until it's done to continue.
Adding a Run until find fails option will solve all of these issues, and as a software developer, I cannot see this breaking existing macros, but adding an extremely valuable feature (in my use cases)

**Comment by Arkaine80** [comment]
> [@Arkaine80](https://github.com/Arkaine80)
> 
> > Adding a Run until find fails option will solve all of these issues
> 
> I don't think you'll ever see this kind of enhancement to the macro subsystem.
> 
> > I start by finding the main tag let's go with "Data":, do a find for that element let's go with Fullname, move right until after the ":, shift End -2 chars, cut, ctrl+home.
> 
> Unfortunately you don't provide enough info on what you're finding for me to provide a more complete example, but if you turn to scripting, you can get started with something like:
> 
> while True:
>     result = editor.findText(FINDOPTION.MATCHCASE, editor.getCurrentPos(), editor.getLength(), 'Data')
>     if result is None:
>         break
>     else:
>         (match_start_pos, match_end_pos) = result
>         result = editor.findText(FINDOPTION.MATCHCASE, match_end_pos, editor.getLength(), 'Fullname')
>         if result is None:
>             break
>         else:
>             # ...
> If you'd like further scripting assistance, please turn to the Community site: https://community.notepad-plus-plus.org/

Thanks for getting back to me. It's a pity you do not see an enhancement like this, I can think of many use cases, but hey, it is what it is.

> Unfortunately you don't provide enough info on what you're finding for me to provide a more complete example, but if you turn to scripting, you can get started with something like:

I want to turn this:

{"Data:"
[
	{
		"UserID": 12345,
		"Fullname": "John Solomon",
		"Surname": "Smith"
	},
	{
		"UserID": 13346,
		"Fullname": "Joe Henry",
		"Surname": "Soap"
	}
]
}

into this:

John Solomon
Joe Henry

**Comment by xomx** [comment]
@Arkaine80 
> It's a pity you do not see an enhancement like this happening

You don't need it, try to record these steps (before you must have the cursor at the very beginning of your file/tab, simply press Ctrl+Home for that if you are somewhere else):
- menu > Macro > Start Recording
- press `Ctrl+Shift+B` (Begin Select)
- press `Ctrl+F`, Find what: `"Fullname": "`, press `Find Next`, close Find dialog
- press `Ctrl+Shift+B` (End Select)
- press `Delete`
- press `Ctrl+F`, Find what: `",`, press `Find Next`, close Find dialog
- press `Delete`
- press the `Right Arrow key` once
- menu > Macro > Stop Recording
- menu > Macro > Run a Macro Multiple Times > select `Run until the end  of file`, click `Run`

You'll end up with this, which can be easily trimmed to your desired final form (just delete the rest of the file):

![Image](https://github.com/user-attachments/assets/dce8f1fe-58fc-42f1-954e-36ea8d699fae)

**Comment by Arkaine80** [comment]
@xomx 

Great workaround, I did not know about Ctrl+Shift+B. Thanks for the assistance, much appreciated.

**Comment by Rekrullurker** [comment]
And what if you want to turn this;

"UserID": 12345,
"Fullname": "John Solomon",
"Surname": "Smith"
},
{
"UserID": 13346,
"Fullname": "Joe Henry",
"Surname": "Soap"

Into this;

"UserID": 12345,
"Fullname": "Solomon, John",
"Surname": "Smith"
},
{
"UserID": 13346,
"Fullname": "Henry, Joe",
"Surname": "Soap"

Easy if the macro only executes on a successful search, however with the current "dumb" macro system, it will corrupt anything after the last match because it will just keep executing the parts of the macro that copy and move part of the line.

The part that amazes me the most, is the amount of hostility and resistance to the idea of making macros just a TINY bit smarter, even though that change wouldn't break any existing macros, and would open up a lot of flexibility in creating macros.

"Hey guys, can we add one little option to make the macros smarter?"

"HELL NO! I WANT MY MACROS TO BE DUMB AS A F***ING POST!!! If you want even the absolute bare minimum of decision making in a macro, go spend a few days learning a scripting language!"

**Comment by xomx** [comment]
> Into this;
> 
> "UserID": 12345, "Fullname": "Solomon, John", "Surname": "Smith" }, { "UserID": 13346, "Fullname": "Henry, Joe", "Surname": "Soap"

Then I have these names reversed and with comma, of course, but my macro works the same as with the previous data (thus it needs the simple deleting of the rest of the file as before):
```
Solomon, John
Henry, Joe
urname": "Soap"
```

> it will corrupt anything after the last match because it will just keep executing

Yes. So if I know about it, then I can adapt, right?

> The part that amazes me the most, is the amount of hostility and resistance to the idea of making macros just a TINY bit smarter

My previous 1st reply did not mean that I would somehow belittle the argument that it would be good to have more control over the macros if they do not find what the user expected to always be in the source data (e.g. in processing slightly malformed or incomplete user-filled data)

> the RANT 

I think we have the usual misunderstanding here of how opensource works. It would certainly be good to have many other improvements in this app, not just the one suggested in this issue. But each contributor here has only limited resources, mainly free time, but also knowledge of the given issue or the relevant code.

I'll try to put it as simply as possible - while there are many "thinkers" here who would know how to do something better, there are very few real "doers" who could actually fulfill their desire. You wrote:

> go spend a few days learning a scripting language!

so you don't even want to spend your precious time learning easier scripting, but you want the N++ devs to do the hard work for you in C++! Nice.

**Comment by Rekrullurker** [comment]
> Then I have these names reversed and with comma, of course, but my macro works the same as with the previous data (thus it needs the simple deleting of the rest of the file as before):

And what if you don't want the rest of the file deleted?

>Yes. So if I know about it, then I can adapt, right?

That relies on the user taking extra steps to safeguard against a puzzling shortcoming, and it would be extremely easy to forget to do this.

>I think we have the usual misunderstanding here of how opensource works. It would certainly be good to have many other improvements in this app, not just the one suggested in this issue. But each contributor here has only limited resources, mainly free time, but also knowledge of the given issue or the relevant code.

>I'll try to put it as simply as possible - while there are many "thinkers" here who would know how to do something better, there are very few real "doers" who could actually fulfill their desire. You wrote:

Yes, I understand, but here's how I look at it;

Ideally, the authors of any piece of software should want to make it as useful and user friendly as possible, within the limits of what's practical. In my humble opinion, this macro behavior is completely contrary to common sense. When you use the search/find function in a text editor, it's typically because you want to make changes to that text. I can't think of a single instance where you would search for something, the text isn't found, and you say to yourself "I'll just make my planned changes right here on this random line, regardless of what it does to the file." Doing that wouldn't make any sense, but yet that's what the macro does.

I'll freely admit that I'm not a programmer. I never learned C, or modern assembly, however I know enough that I'm confident it would be a relatively simple change for anyone familiar with the code to make. Test to see if the search failed, and if so, end the macro. Adding a checkbox to make this change optional would require a little more work, but honestly, I can't see why anyone would want to preserve the current behavior.

>so you don't even want to spend your precious time learning easier scripting, but you want the N++ devs to do the hard work for you in C++! Nice.

They're more familiar with their code, than I am with any of the mentioned scripting languages. The only scripting language that I'm familiar with is Windows Batch, and even that trips me up all the time with its various quirks. Not to mention that I see examples all the time that I don't understand at all.

And there is a vast percentage of computer users today for whom learning scripting of any type is completely beyond their ability. Most people today have gotten so used to software offering to do everything for them, that they often don't know how to do anything outside that. 

Which is why making the macros as intuitive as possible seems like a good idea to me.

To be perfectly honest, when I first created this thread, I thought that this was a bug and that the authors would be glad I reported it. When I learned that it wasn't a bug, I thought people would be like "Yeah, having it be conditional makes a lot of sense." I never thought that people would defend a program being less functional.

**Comment by alankilborn** [comment]
>I know enough that I'm confident it would be a relatively simple change for anyone familiar with the code to make

I had some very recent thoughts upon this topic [HERE](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/16437#issuecomment-2827278358) if anyone cares to read.

>I thought that this was a bug and that the authors would be glad I reported it.

Sure, glad to have it.

>I never thought that people would defend a program being less functional.

I don't think anyone is doing that.

---

The issue can remain open and if someone comes along and wants to work on it, then I say "awesome".

**Comment by xomx** [comment]
@Rekrullurker 

@alankilborn said it all for me, but can't help myself and comment on at least this:

> I'm not a programmer. I never learned C, or modern assembly, however I know enough that I'm confident it would be a relatively simple change for anyone familiar with the code

THIS!
How many times an in how many forms have I heard something like this! (not only here, also IRL).

> I never thought that people would defend a program being less functional.

I see we really don't understand each other. I would also appreciate it if someone would take care of and implement here a seamless extension to the existing macro control.

**Comment by victorel-petrovich** [comment]
@Rekrullurker  in https://github.com/notepad-plus-plus/notepad-plus-plus/issues/4809#issuecomment-417377954  above: 
> Then what's the point of being able to include a search in a macro if it doesn't matter whether the search succeeds or not?> ...
> Can you please provide an example of a situation where you would want to use search (not search & replace) in a macro, but **NOT** have the rest of the macro be conditional on the search succeeding?

I can: when you know for sure that what you search for is there (because you see it). But it's still useful to use the Find function in order to get that sequence of chars selected (to save you from manual selection) - and then the rest of your macro can work on it. (I could give the specific example if you don't believe). 

@ghost in https://github.com/notepad-plus-plus/notepad-plus-plus/issues/4809#issuecomment-417363198 above and next ones: 
> Imho a macro is dump it takes what it got and executes it.
If you need logic then you need some kind of scripting language 
>...
> If you want to have a macro to behave more intelligent, than you have to define its behavior which means
you need to describe how to use it and under which circumstances - in code.
>...
> I understand the desire to have more control over macros but I do not know how to do that without changing the whole macro concept.
How should the logic decide whether, for example, finding a string is an error or not.
> ...
> To realize such a thing it needs, in my opinion, a kind of flow control which would mean,
for example, while recording with each step it can be defined whether the next step, in the case of an error, should be executed or not. Sounds very much like a scripting language to me.

If you All don't mind **_an exercise in imagination_**... 
I think rudimentary flow control could _in principle_ be added: a few "blocks" menu entries in the Menu>Macros and/or in the toolbar Macro group for: 
- conditional execution: `IF` and, better, also `ELSE` blocks,  that test a condition and execute one or the other sequence of macro actions depending on the test result. 
- conditional repetition: `WHILE` block, that test a condition and repeat a sequence   of macro actions if it's true
- an `END` block to go along with the others above

Both kinds of conditionals should allow access to some built-in variables for building the conditions in the above 2 blocks; in particular, variables that hold the status of last Find. The more such variables accessible, the more power the macros will have. 

Possible workflow:
- You press the Record button/icon
- perform actions A, B, C 
- click on the `IF` macro icon , which opens a dialog to prompt you for a condition (like a choice of variable to test, a choice of a comparison operator , and right operand - say an integer) 
- perform actions D, E, F (corresponding to the "true" branch), 
- press on the `ELSE` icon, 
- perform actions G, H (corresponding to the "false"branch), 
- click on the `END` block/icon
- optionally perform a few more actions I,J,K, 
- press Stop recording icon. 

I know many of you would think this is a stretch and unconventional for macros, but would be the ultimate generalization of what Notepad++ already has in Macro > Run a macro multiple times... > options to select N times or till the end of the file, 
and more in other editors as @deadeyelo mentioned in https://github.com/notepad-plus-plus/notepad-plus-plus/issues/4809#issuecomment-453795737 above: 
> Of the several editors that support macros that I have worked with, Windows, OpenVMS, they abort macro execution on some errors failed search, down arrow, up arrow when at end or beginning of buffer, execution for more iterations than exist in file (the Run Multiple Times, executing 10 times when only 5 times is possible), etc. 

@xomx 
> I'll try to put it as simply as possible - while there are many "thinkers" here who would know how to do something better, there are very few real "doers" who could actually fulfill their desire.  

~~Cannot argue with that.~~ At first read seems true, but the category of "doers" is small not necessarily for the lack of ability.

**Comment by Rekrullurker** [comment]
Not sure why I never saw this;

> THIS!
> How many times and in how many forms have I heard something like this! (not only here, also IRL).

All coding, no matter what language its in, is based on the same basic principles of variables, comparisons, branches, subroutines, etc. Do you personally know 6502 assembly? If I were to tell you that it's impossible to create a loop that counts to a value higher than 255 in 6502 assembly, would you assume that 255 is the highest number an 8-bit computer from the 80s could ever count to, or would you believe that there's a simple work-around, despite not knowing the language?

When Notepad++ performs a search, it has to set a flag based on whether the search succeeded or failed because it has to know whether it needs to update the cursor position to the location of the text that was found, or not. And if it was a success, that also means that at the very least, the pointers to the current cursor position, as well as the window position within the file have to be changed. 

To the user, macros may just be a series of recorded functions, but the program needs to interpret each part of the macro as code because it needs to be able to jump to the relevant part of the code to perform the requested action. If the user performs a search in the macro, the program has to know to insert the search text into the appropriate variable and then jump to the subroutine that actually performs the search. So what's needed is to add a check at the point where it returns from performing the search, but before the rest of the macro executes, that will check the status of the success/fail flag for that search, and if it's a fail, end the macro execution. 

Is any of the above incorrect? Does the fact that I don't know how to do this in C invalidate the above idea? If so, please explain to me why adding a simple flag check after a search is performed, would be too difficult. And if the argument is that some third party needs to come along and be willing to contribute that particular portion of the code, I have to wonder about the authors and current contributors, that they are unable to make such a change despite being arguably more familiar with the code than someone who just shows up out of the blue with an idea.

> I can: when you know for sure that what you search for is there (because you see it). But it's still useful to use the Find function in order to get that sequence of chars selected (to save you from manual selection) - and then the rest of your macro can work on it. (I could give the specific example if you don't believe).

No. I asked for an example of using search in a macro, but where it doesn't matter if the search succeeds or fails. By manually confirming that the text exists before executing the macro with the search, you're ensuring that it _will_ succeed. Actually, you've just supported my original position that the macro should only execute if the text is found, because in your example, you wouldn't be using that macro if you didn't already know that the text was there.

How about this:

Can anyone give me an example of a way in which it would _break_ things if it were changed to make macro execution conditional on a search succeeding? To be clear, I want someone to post an example of a situation where you want the macro to search for something, _fail_ to find that text, but continue executing anyway, and if it were changed to stop executing when the search fails, that would cause a problem. And also please explain why, if it doesn't matter if the search fails, why are you even including it in the macro? What purpose does it serve if its success or failure is completely irrelevant to the macro's execution?

**Comment by victorel-petrovich** [comment]
@Rekrullurker , the reason I replied with that explanation of an example is that, in my quotation of you and in a few other sentences, you sounded (to me) like you didn't know Search in a macro CAN be useful even though the rest of the macro is NOT conditional on the Search status. For that "rest of macro" to be conditional in _technical_ sense, it means that it has to check that status flag and conditionally execute or not.

FWIW , I agree that it would be more useful to make that execution  _somehow_ conditional, 
BUT I think it's one-sided/ biased to insist on choosing, for everyone, that the default should be this rule:

**R1.** execute the rest of the macro if and only if the search succeeded

as opposed to this

**R2.** execute the rest of the macro if and only if the search failed. 

Which was mentioned in earlier comments by a @ghost: 
> In this context, assuming that one wants to execute the macro in documents which do NOT contain
a particular text is as valid as one wants to run the macro if it HAS the text looking for.
How could we know beforehand?

And so, that is why, to make everyone happy, I am for a more powerful, and general, conditional operator, of the kind I described in https://github.com/notepad-plus-plus/notepad-plus-plus/issues/4809#issuecomment-3239606439 . 

Additionally, It's not hard to imagine that there are other actions also having some sort of success status, and based on which users will also want to request a conditional execution feature. 
There will be logic, and thus code, that will overlap in all these separate conditional features, that I believe best to make into a conditional that can be used for any user-supplied condition and status flags or even other kinds of N++ variables.

**Comment by Rekrullurker** [comment]
If it were changed to make the rest of the macro conditional on the search succeeding, it would have no detrimental effect on the example you gave. I understand that not everyone wants the same thing, however I am having trouble thinking of a single case where making it conditional would cause any problems for anyone. 

Let's say that someone is following your example, and they know for a fact that the file contains the word "TEST", so they create this macro;

1. Search for TEST
2. Delete from that word to the end of the line

As of right now, they would need to know how many times TEST appears in the file, and either manually execute the macro that many times, or tell it to repeat that many times. If the macro stopped on a search failure, it would work exactly the same, with the exception that it wouldn't make any changes to the file if the user was mistaken and the text wasn't in the file, or they had miscounted how many times it occurred.

As it stands now, it's not even possible to make a macro that only works if the search fails, as the macro doesn't pay any attention to the search whatsoever. It performs the search and then continues regardless of the results.

I also understand that people might want other actions to be conditional on a variety of factors, but I have singled out search because because without the macro being conditional, search is only useful in a macro if you are absolutely certain that the text appears in the file and how many times. Beyond that, it can be actively dangerous to use, if the file _doesn't_ contain the text. 

A little backstory...

My text editor of choice was always TextPad 4. I noticed that the spell checker seemed to be making some odd, automatic changes to the files I was using it on, so I went looking for something else, which lead me to try Notepad++. I've written a bunch of macros in TextPad, and wanted to see if Notepad++ could match it.

Here's a concrete example: I created a macro that goes through a list of movie titles searching for the word "The", and when it finds it, it selects it, deletes it and the space after it, moves to the end of the line, adds a comma, then pastes in "The", jumps back to the start of the line, then drops down a line. When executed to the end of the file, it will automatically change this;

The Terminator
The Day the Earth Stood Still
The Time Machine

To this;

Terminator, The
Day the Earth Stood Still, The
Time Machine, The

Then a second macro selects all, and sorts them in alphabetical order.

I can run the macro on any list of movies, confident that it will only change the ones that have the word "The" in front of them. If I had a list of 100+ titles and had no idea how many of them started with "The", it would be more work with Notepad++. I would first have to sort the list to group all the movies with "The" together, scroll down to the "T"s, select those titles manually, then execute the macro through the selection. It somewhat negates the convenience of making macros when you have to babysit them.
